runelite-client/src/main/java/net/runelite/client/plugins/inferno/InfernoNPC.java

```diff
@@ -25,162 +25,472 @@
 package net.runelite.client.plugins.inferno;
 
 import java.awt.Color;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
 import lombok.AccessLevel;
 import lombok.Getter;
-import lombok.Setter;
+import net.runelite.api.AnimationID;
+import net.runelite.api.Client;
 import net.runelite.api.NPC;
 import net.runelite.api.NpcID;
+import net.runelite.api.Prayer;
+import net.runelite.api.coords.WorldArea;
+import net.runelite.api.coords.WorldPoint;
+import org.apache.commons.lang3.ArrayUtils;
 
 public class InfernoNPC
 {
-	public enum Attackstyle
+	@Getter(AccessLevel.PACKAGE)
+	private NPC npc;
+	@Getter(AccessLevel.PACKAGE)
+	private Type type;
+	@Getter(AccessLevel.PACKAGE)
+	private Attack nextAttack;
+	@Getter(AccessLevel.PACKAGE)
+	private int ticksTillNextAttack;
+	private int lastAnimation = -1;
+	private boolean lastCanAttack;
+	//0 = not in LOS, 1 = in LOS after move, 2 = in LOS
+	private HashMap<WorldPoint, Integer> safeSpotCache;
+
+	InfernoNPC(NPC npc)
 	{
-		MAGE("Mage", Color.CYAN),
-		RANGE("Range", Color.GREEN),
-		MELEE("Melee", Color.WHITE),
-		RANDOM("Random", Color.ORANGE);
+		this.npc = npc;
+		this.type = Type.typeFromId(npc.getId());
+		this.nextAttack = type.getDefaultAttack();
+		this.ticksTillNextAttack = 0;
+		this.lastAnimation = -1;
+		this.lastCanAttack = false;
+		this.safeSpotCache = new HashMap();
+	}
 
-		@Getter(AccessLevel.PACKAGE)
-		private String name;
+	void updateNextAttack(Attack nextAttack, int ticksTillNextAttack)
+	{
+		this.nextAttack = nextAttack;
+		this.ticksTillNextAttack = ticksTillNextAttack;
+	}
 
-		@Getter(AccessLevel.PACKAGE)
-		private Color color;
+	void updateNextAttack(Attack nextAttack)
+	{
+		this.nextAttack = nextAttack;
+	}
 
-		Attackstyle(String s, Color c)
+	boolean canAttack(Client client, WorldPoint target)
+	{
+		if (safeSpotCache.containsKey(target))
+		{
+			return safeSpotCache.get(target) == 2;
+		}
+
+		boolean hasLos = new WorldArea(target, 1, 1).hasLineOfSightTo(client, this.getNpc().getWorldArea());
+		boolean hasRange = this.getType().getDefaultAttack() == Attack.MELEE ? this.getNpc().getWorldArea().isInMeleeDistance(target)
```

Comment:
Is this correct? Players can still attack instantly when up to 2 squares out of their range, as the first attack will happen in the same time as walking towards the opponent. I'm not sure if this is the case for monsters though

Reply:
I'm not entirely sure, but we tried with a variety of values and these seemed to give the most consistent and correct results. It's also how it was done in the previous inferno plugin. I'll keep this in mind and make adjustments if necessary.